src/providers/structure.ts

```diff
@@ -90,152 +74,62 @@ export class SectionNodeProvider implements vscode.TreeDataProvider<Section> {
 
-        const pathRegexp = new PathRegExp()
-        const headingReg = RegExp(pattern, 'm')
         const envNames = this.showFloats ? ['figure', 'frame', 'table'] : ['frame']
-        const envReg = RegExp(`(?:\\\\(begin|end)(?:\\[[^[\\]]*\\])?){(?:(${envNames.join('|')})\\*?)}`, 'm')
-        const labelReg = /\\label{([^}]*)}/m
 
         const lines = content.split('\n')
         for (let lineNumber = 0; lineNumber < lines.length; lineNumber++) {
             const line = lines[lineNumber]
-            pathRegexp.resetLastIndex()
-            let result = envReg.exec(line)
-            if (result && result[1] === 'begin') {
-                envStack.push({name: result[2], start: lineNumber, end: lineNumber})
-                continue
-            } else if (result && result[2] === envStack[envStack.length - 1].name) {
-                const env = envStack.pop()
-                if (!env) {
-                    continue
-                }
-                env.end = lineNumber
-                const caption = this.getCaptionOrTitle(lines, env)
-                if (!caption) {
-                    continue
-                }
-                const depth = noRoot() ? 0 : currentRoot().depth + 1
-                const newEnv = new Section(`${env.name.charAt(0).toUpperCase() + env.name.slice(1)}: ${caption}`, vscode.TreeItemCollapsibleState.Expanded, depth, env.start, env.end, filePath)
-                if (noRoot()) {
-                    children.push(newEnv)
-                } else {
-                    currentRoot().children.push(newEnv)
-                }
-                continue
-            }
+            // Environment part
+            structureModel.buildEnvModel(envNames, lines, lineNumber)
 
-            // inputs part
+            // Inputs part
             if (imports) {
-                const matchPath: MatchPath | undefined = pathRegexp.exec(line)
-                if (matchPath) {
-                    // zoom into this file
-                    // resolve the path
-                    const inputFilePath: string | undefined = pathRegexp.parseInputFilePath(matchPath, filePath, this.extension.manager.rootFile ? this.extension.manager.rootFile : filePath)
-                    if (!inputFilePath) {
-                        this.extension.logger.addLogMessage(`Could not resolve included file ${inputFilePath}`)
-                        continue
-                    }
-                    // Avoid circular inclusion
-                    if (inputFilePath === filePath || newFileStack.includes(inputFilePath)) {
-                        continue
-                    }
-                    if (prevSection) {
-                        prevSection.subfiles.push(inputFilePath)
-                    }
+               const inputFilePath = structureModel.buildImportModel(line)
+               if (inputFilePath) {
                     this.buildModel(inputFilePath, newFileStack, rootStack, children, sectionNumber)
```

Comment:
Is there possibility that this line results in an infinite loop?

Reply:
It could happen and in this case the promise returned by `provideDocumentSymbols` will fail. I am not sure how far we should go to properly handle the infinite loop error as it basically means that either there is an error in this file or it is a false circular inclusion because we de not interpret LaTeX code.

Reply:
The following is an example I have in mind to avoid infinite loops.

- https://github.com/tamuratak/LaTeX-Workshop/tree/use_memo
- https://github.com/James-Yu/LaTeX-Workshop/commit/6a36a50b38cdd932a007b95230d3e2f4cf6303d7

Reply:
That is indeed the simplest way of doing it. We already use this pattern quite a lot in LW.